File: apps/relay-server/src/realtime/first-message.ts
import type { SessionManager } from './session-manager.js';
import type { Language, SessionMode } from '../types.js';
import { firstMessageTemplates, firstMessageAgentTemplates } from '../prompt/templates.js';

interface FirstMessageConfig {
  mode: SessionMode;
  targetLanguage: Language;
  service?: string;
  recipientDetectionTimeoutMs?: number;
}

/**
 * First Message Strategy (C-3):
 * 1. Wait for recipient to answer (Session B detects speech via Server VAD)
 * 2. Send AI disclosure + greeting via Session A
 * 3. After greeting, enable user input (Relay Mode) or start autonomous conversation (Agent Mode)
 */
export class FirstMessageHandler {
  private recipientDetected = false;
  private firstMessageSent = false;
  private detectionTimeout: ReturnType<typeof setTimeout> | null = null;
  private onComplete: (() => void) | null = null;
  private onTimeout: (() => void) | null = null;

  constructor(
    private sessionManager: SessionManager,
    private config: FirstMessageConfig,
  ) {}

  /**
   * Start waiting for recipient to answer.
   * Returns a promise that resolves when first message sequence is complete.
   */
  start(callbacks: {
    onComplete: () => void;
    onTimeout: () => void;
  }) {
    this.onComplete = callbacks.onComplete;
    this.onTimeout = callbacks.onTimeout;

    // Set timeout for no answer (15 seconds default)
    const timeoutMs = this.config.recipientDetectionTimeoutMs ?? 15_000;
    this.detectionTimeout = setTimeout(() => {
      if (!this.recipientDetected) {
        console.log('[FirstMessage] Recipient detection timeout');
        this.onTimeout?.();
      }
    }, timeoutMs);
  }

  /**
   * Called when Session B detects recipient's first speech (Server VAD).
   */
  handleRecipientDetected() {
    if (this.recipientDetected) return;

    this.recipientDetected = true;
    if (this.detectionTimeout) {
      clearTimeout(this.detectionTimeout);
      this.detectionTimeout = null;
    }

    console.log('[FirstMessage] Recipient detected, sending AI disclosure');
    this.sendFirstMessage();
  }

  /**
   * Called when Session A completes first message TTS.
   */
  handleFirstMessageComplete() {
    if (this.firstMessageSent) return;
    this.firstMessageSent = true;

    console.log('[FirstMessage] First message complete');
    this.onComplete?.();
  }

  private sendFirstMessage() {
    const { mode, targetLanguage, service } = this.config;

    let message: string;

    if (mode === 'agent') {
      message = firstMessageAgentTemplates[targetLanguage]
        .replace('{{service}}', service ?? 'your inquiry');
    } else {
      message = firstMessageTemplates[targetLanguage];
    }

    this.sessionManager.triggerFirstMessage(message);
  }

  cleanup() {
    if (this.detectionTimeout) {
      clearTimeout(this.detectionTimeout);
      this.detectionTimeout = null;
    }
  }
}

File: apps/relay-server/src/realtime/first_message.py
"""First Message Strategy — AI 고지 + 수신자 인사 대기.

PRD 3.4:
  1. Twilio가 전화를 건다
  2. 수신자가 전화를 받는다 ("여보세요")
  3. Session B가 수신자 첫 발화를 감지 (Server VAD)
  4. 자동 AI 고지 (Session A → Twilio → 수신자)
  5. AI 고지 완료 후:
     - Relay Mode: User 앱에 "상대방이 응답했습니다" 알림
     - Agent Mode: AI가 바로 용건 시작
"""

import logging
from typing import Callable, Coroutine

from src.prompt.templates import FIRST_MESSAGE_TEMPLATES
from src.realtime.session_a import SessionAHandler
from src.types import ActiveCall, CallMode, CallStatus, WsMessage, WsMessageType

logger = logging.getLogger(__name__)


class FirstMessageHandler:
    """수신자의 첫 발화를 감지하고 AI 고지를 트리거한다."""

    def __init__(
        self,
        call: ActiveCall,
        session_a: SessionAHandler,
        on_notify_app: Callable[[WsMessage], Coroutine],
    ):
        self.call = call
        self.session_a = session_a
        self._on_notify_app = on_notify_app

    async def on_recipient_speech_detected(self) -> None:
        """수신자의 첫 발화가 감지되면 AI 고지를 전송한다."""
        if self.call.first_message_sent:
            return

        self.call.first_message_sent = True
        self.call.status = CallStatus.CONNECTED

        logger.info("Recipient answered — sending AI greeting (call=%s)", self.call.call_id)

        # Session A가 이미 응답 중이면 대기 (conversation_already_has_active_response 방지)
        if self.session_a.is_generating:
            logger.debug("Waiting for Session A to finish before sending greeting...")
            await self.session_a.wait_for_done(timeout=3.0)

        # AI 고지 메시지 전송 (Session A → Twilio → 수신자)
        # 영어 원문을 Session A에 전달 → Session A가 target_language로 번역
        greeting = FIRST_MESSAGE_TEMPLATES.get(
            self.call.target_language,
            FIRST_MESSAGE_TEMPLATES["en"],
        )
        wrapped = f"[User says in {self.call.source_language}]: {greeting}"
        await self.session_a.send_user_text(wrapped)

        # App에 통화 연결 알림
        if self.call.mode == CallMode.RELAY:
            await self._on_notify_app(
                WsMessage(
                    type=WsMessageType.CALL_STATUS,
                    data={
                        "status": "connected",
                        "message": "상대방이 응답했습니다. 말씀하세요.",
                    },
                )
            )
        else:
            await self._on_notify_app(
                WsMessage(
                    type=WsMessageType.CALL_STATUS,
                    data={
                        "status": "connected",
                        "message": "상대방이 응답했습니다. AI가 대화를 시작합니다.",
                    },
                )
            )
